include/Server/include/ClientSession/src/ClientSession.cpp

```diff
@@ -0,0 +1,768 @@
+                m_socket.WriteAsync(SmtpResponse(SmtpResponseCode::OK).ToString()).get();
+            }
+
+        }
+        catch (const boost::system::system_error& e)
+        {
+            // Handle specific boost system errors (e.g., client disconnection)
+            Logger::LogError("System error in ClientSession::HandleData(Client disconnected): " + std::string(e.what()));
+        }
+        catch (const std::exception& e)
+        {
+            // Handle other exceptions
+            Logger::LogError("Exception in ClientSession::HandleData: " + std::string(e.what()));
+            throw;
+        }
+
+        Logger::LogProd("Data handling complete.");
+    }
+    catch (const std::exception& e)
+    {
+        Logger::LogError("Exception in ClientSession::HandleData: " + std::string(e.what()));
+    }
+
+    Logger::LogDebug("Exiting ClientSession::HandleData");
+}
+
+void ClientSession::HandleRset(const SmtpRequest& request)
+{
+    Logger::LogDebug("Entering ClientSession::HandleRset");
+    Logger::LogTrace("ClientSession::HandleRset parameter: const SmtpRequest reference: " + request.data);
+
+    try
+    {
+        m_mail_builder = MailMessageBuilder();
+        m_socket.WriteAsync(SmtpResponse(SmtpResponseCode::OK).ToString()).get();
+        m_current_state = ClientState::AUTH_SENT;
+    }
+    catch (const std::exception& e)
+    {
+        Logger::LogError("Exception in ClientSession::HandleRset: " + std::string(e.what()));
+    }
+
+    Logger::LogDebug("Exiting ClientSession::HandleRset");
+}
+
+/******************/
+/* State handlers */
+/******************/
+void ClientSession::HandleConnectedState(const SmtpRequest& request)
+{
+    Logger::LogDebug("Entering ClientSession::HandleConnectedState");
+    Logger::LogTrace("HandleConnectedState::HandleEhlo parameter: const SmtRequest reference");
+    
+    try
+    {
+        if (request.command == SmtpCommand::EHLO)
+        {
+            m_socket.WriteAsync(SmtpResponse(SmtpResponseCode::OK).ToString()).get();
+            m_current_state = ClientState::EHLO_SENT;
+        }
+        else 
+        {
+            Logger::LogError(std::to_string(static_cast<int>(m_current_state)));
+            m_socket.WriteAsync(SmtpResponse(SmtpResponseCode::BAD_SEQUENCE).ToString()).get();
+        }
+    }
+    catch (const std::exception& e)
+    {
+        Logger::LogError("Exception in ClientSession::HandleConnectedState: " + std::string(e.what()));
+    }
+
+    Logger::LogDebug("Exiting ClientSession::HandleConnectedState");
+}
+
+void ClientSession::HandleEhloSentState(const SmtpRequest& request)
+{
+    Logger::LogDebug("Entering ClientSession::HandleEhloSentState");
+    Logger::LogTrace("ClientSession::HandleEhloSentState parameter: const SmtpRequest reference: " + request.data);
+
+    if (request.command == SmtpCommand::STARTTLS)
+    {
+        HadleStartTls(request);
+        m_current_state = ClientState::STARTTLS_SENT;
+    }
+    else
```

Comment:
After EHLO command client does not necessarily send STARTTLS. Client can even do not send STARTTLS at all